File: LeetCode/Medium/0443-string-compression/0443-string-compression.py
from collections import deque


class Solution:
    def compress(self, chars: List[str]) -> int:
        stack = deque(chars)
        compressed = deque()
        while stack:
            c = stack.popleft()
            compressed.append(c)

            cnt = 1
            while stack and stack[0] == c:
                cnt += 1
                stack.popleft()

            if cnt == 1:
                continue

            compressed.extend(str(cnt))

        chars[:] = list(compressed)
        return len(chars)

File: LeetCode/Medium/0443-string-compression/0443-string-compression.rs
use std::collections::VecDeque;

impl Solution {
    pub fn compress(chars: &mut Vec<char>) -> i32 {
        let mut stack = VecDeque::from(chars.clone());
        let mut compressed: VecDeque<char> = VecDeque::new();
        while !stack.is_empty() {
            compressed.push_back(stack.pop_front().unwrap());

            let mut count = 1;
            while !stack.is_empty() && stack[0] == compressed[compressed.len() - 1] {
                count += 1;
                stack.pop_front();
            }

            if count == 1 {
                continue;
            }

            compressed.extend(count.to_string().chars().collect::<Vec<char>>());
        }

        chars.clear();
        chars.extend(compressed);

        return chars.len() as i32;
    }
}
